ELF support roadmap (my homework from the Linux training day)

trylek opened the pull request.

trylek committed:
ELF support roadmap (my homework from the Linux training day)

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,170 @@
+  has been located. Fully switching over to native ELF format for ready-to-run
+  binaries is more complicated as it involves not yet implemented embedded MSIL
+  in native R2R binaries, a new concept we're currently developing as part of
+  the composite R2R effort; this will replace the existing lookup for COR header
```
There is way too many places in the ecosystem that depend on the IL .dll files on disk. I do not think it will replace it. It may be one more option.

trylek replied:
Are you suggesting any specific change to the wording here?

jkotas replied:
I would delete the whole sentence "this will replace the existing lookup for COR header using the COM descriptor directory entry in the PE header"

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,170 @@
+  decoupled from the PE format by some form of virtualization. We need to
+  port the various ELF header structures to managed code for use in Crossgen2
+  and R2RDump (do we need some approval or special arrangements for porting
+  the Linux header files like elf.h and elftypes.h to managed code in the runtime
```
This is documented in https://github.com/dotnet/runtime/blob/master/CONTRIBUTING.md#copying-files-from-other-projects

We have been using these headers from LLVM that have compatible license: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/inc/llvm/ELF.h

jkotas replied:
So do not talk about these as "Linux header files" because they are unlikely to come from Linux due to incompatible license.

trylek replied:
Fixed in 3rd commit, thanks for clarifying!

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,170 @@
+  should be pretty straightforward in terms of just filling in the various header
+  and section tables and emitting the section data in the proper place.
+
+* **Diagnostics / SOS**: switching over to native ELF will have fundamental effect on
```
I do not see why switching to ELF envelope for R2R has fundamental effect on SOS

trylek replied:
OK, JanV also mentioned offline some of these might be no-concerns.

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,170 @@
+  for a hard-coded ELF file used as a resource by the test.
+  
+  An interesting observation is that most ELF header structures have
+  paraller 32-bit and 64-bit variants. It might be useful to logically unify them
```
```suggestion
  parallel 32-bit and 64-bit variants. It might be useful to logically unify them
```

trylek replied:
Thanks, I actually noticed another typo, will fix both immediately

trylek replied:
OK, I cannot find "the other typo" right now. I fixed this one in the 2nd commit.

trylek committed:
Fix typo spotted by JanK (paraller -> parallel)

I'm sure I saw another typo but I cannot find it just now. Will
likely send out an update later.

davidwrighton commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,170 @@
+  it will take time to bridge the user experience gap. It is likely that for some
+  temporary period of time we'll need to keep support for both PE files and ELF
+  files on Linux until all diagnostic functionality for ELF has been put on par
+  with pre-existing diagnostic support for PE-based executables.
```
The biggest place I see impact on the diagnostic tools is on profiling, and possibly native debugging. Right now we use crossgen to generate a native pdb to interface with the profiler, but with the generation of ELF files directly, I believe we would also want to at least cost out the effort of generating DWARF debugging information for these files. Providing a means for pushing some degree of managed debugging into the native Unix debugging stack would be quite valuable, and potentially reduce the problems caused by a lack of support for mixed mode debugging in managed code. But, as Jan notes, these aren't concerns about modifying existing tools so much as enabling entire new categories of capability.

trylek replied:
Thanks David for pointing that out, I believe this should be definitely somehow made part of this design doc even though the actual prioritization is yet to be determined.

noahfalk replied:
> cost out the effort of generating DWARF debugging information

For debugging it could make some difference, profiling I'm not sure it would. On Linux we generate perfmaps for the 'perf' tool already and that I'm not sure if we lack anything that DWARF would give us instead. For Visual Studio profiler resolving R2R image symbols doesn't currently work - we are discussing getting the symbolic data into the trace at rundown rather than out-of-line tools to create symbol formats.

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,170 @@
+  the choice of PE / ELF format and use the newly created ELF API to [re]construct
+  the necessary headers and create a fully functional ELF executable.
+
+  At this point this will probably still be just a dynamic library, not a
```
For this, crossgen needs to emit `.o` file and the final binary needs to be produced by platform linker (ie `ld` on Linux).

trylek replied:
OK, I guess this means we basically need to be able to generate an object or executable flavor of the ELF file. At least I hope they differ less than obj and exe on Windows :-).

swaroop-sridhar replied:
Generating `.o` from crossgen and linking later would be very useful feature -- this will enable building native linked single-exe apps. 

However, how will cross compilation of apps work? We'll need cross-platform native tool chains (ex: `ld` on Windows)

trylek committed:
Clarifications and corrections per JanK's PR feedback

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+  from the existing Linux libraries. We should consider some level of
+  virtualization to cater for common constructs. The new objects should be
+  Linux-specific and should use the Linux-native functions for manipulating
+  executables like **dlopen** / **dlsym**.
```
Nit: The markdown convention is to quote identifiers using backticks

(multiple places)

trylek replied:
Fixed even though this is actually the only place I found - are you aware of others?

trylek committed:
Backticks

jkotas commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+verifiable / diagnosable is as follows:
+
+* Import (rewrite) the relevant ELF headers to managed code, probably
+  somewhere under *src/coreclr/src/tools/Common/Internal/Runtime*.
```
This can use backticks as well.

trylek replied:
Fixed in 5th commit.

trylek committed:
More backticks per JanK's suggestions

am11 commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+# Introduction
+
+This document details the steps needed for supporting native Linux executable
+format (ELF) for ready-to-run binaries on Linux. Today, ready-to-run binaries
```
Since ELF is supported by non Linux-based as well, could it be kept neutral from that perspective, e.g.:
```suggestion
This document details the steps needed for supporting native executable and linkable
format (ELF) for ready-to-run binaries on Unix-like operating systems. Today, ready-to-run binaries
```
or something that implies `systems that support ELF` (e.g. macOS uses `MACH-O`).

noahfalk commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+(Windows, Linux, OSX). Due to this fact ready-to-run binaries are "unnatural"
+on non-Windows platforms incurring both performance and user experience costs
+(incomplete integration with OS-native application management, inability to
+use OS executable loader, lack of OS-native diagnostic experience). This
```
There are different gradiations of 'OS-native' diagnostics. I am guessing we are proposing only a very scoped set of improvements at this point? An example progression given increasing levels of effort:
1. Use ELF format - tools such as readelf and basic module enumeration in debuggers work
2. Add ELF dynamic symbol table - debuggers/profilers should be able to resolve IPs in those methods to names
3. Add DWARF information for all methods + line information - debuggers/profilers can resolve all pre-compiled methods to names and source, jitted code is still opaque
4. .NET type information encoded with DWARF symbols - Limited object inspection works. Functionality that isn't C-like still probably doesn't work such as variable length types (string/array) or generic instantiations not known at image compile time. 
5. .Net specific DWARF extensions + 1st class support for .Net built into standard tools - Everything works : )

lpereira replied:
It's a shame that the GCC LTO information isn't stable and depends heavily on GIMPLE (GCC's IL); otherwise Crossgen2 could emit it and we could get LTO "for free". (Don't know about Clang's approach here, but I would imagine it depends heavily on the LLVM IR.)

jkotas replied:
> GCC LTO information

A few problems with it:
- The intermediate representation of C/C++ compilers is not aware of GC and it is missing other constructs required to support managed code well. Same problems exist with clang.
- GCC has incompatible license

noahfalk commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+  files on Linux until all diagnostic functionality for ELF has been put on par
+  with pre-existing diagnostic support for PE-based executables.
+
+* **Visual Studio, VS Code**: switching over the executable format is certain to
```
Windbg is also important here if we plan to use these images in Azure. Windbg gets used internally in Azure Watson to debug Linux dumps and we are working to improve its ability to handle managed code.

noahfalk reviewed (commented):
Thanks for writing this up @trylek!

MichalStrehovsky commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+  At some point we may consider porting it to libraries but initially it's better
+  to keep that separate.
+  
+  We should also double-check whether mono doesn't already have something in this
```
Haven't looked at this in detail, but this came up on my Twitter feed recently: https://github.com/xoofx/LibObjectFile. It also comes with some level of DWARF support.

swaroop-sridhar commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+  
+  We should also double-check whether mono doesn't already have something in this
+  vein we could reuse or leverage. We basically need to be able to read / write
+  the main ELF header, the program headers, the section headers and the symbols.
```
Eventually, the ELF R2R assemblies will be loaded via dlopen() on Linux, right? 
It is still necessary to be able to load the assemblies by mapping the individual sections in order to support single-exe apps.

swaroop-sridhar reviewed (approved):
👍

jkotas commented:
> We'll need cross-platform native tool chains (ex: ld on Windows)

Yep. Also, you need the .lib files of the dependencies to feed to `ld`. Producing proper native binaries requires proper toolchain. Shortcuts won't work.

lpereira commented on docs/design/features/readytorun-elf-support-roadmap.md:
```diff
@@ -0,0 +1,166 @@
+  the bitness.
+  
+  We probably don't need to support ELF to the fullest, e.g. endianness doesn't
+  matter as our JIT and runtime doesn't support any high-endian CPU's (ARM has
```
```suggestion
  matter as our JIT and runtime doesn't support any big-endian CPU's (ARM has
```

stephentoub commented:
@trylek, sould this be merged? Closed?

trylek commented:
Closing for now as we're only in the initial planning stages and the doc needs more time to evolve and mature. About to be resurrected in a couple of months I guess.
